docs/digits_diffusion_model.md

```diff
@@ -0,0 +1,822 @@
+- Adaptive noise generation
+
+```{code-cell}
+---
+colab:
+  base_uri: https://localhost:8080/
+id: ZnQqHCAoVfi1
+outputId: a105e2de-ba88-44d0-bad5-3a9a69e54826
+---
+# Initialize training metrics
+losses: List[float] = []                # Store EMA loss history
+moving_avg_loss: Optional[float] = None  # EMA of the loss value
+beta: float = 0.99                      # EMA decay factor for loss smoothing
+
+for epoch in range(num_epochs + 1):
+    # Split PRNG key for independent random operations
+    key, subkey1 = jax.random.split(key)
+    key, subkey2 = jax.random.split(key)
+
+    # Progressive timestep sampling - weights early steps more heavily as training progresses
+    # This helps model focus on fine details in later epochs while maintaining stability
+    progress = epoch / num_epochs
+    t_weights = jnp.linspace(1.0, 0.1 * (1.0 - progress), num_steps)
+    t = jax.random.choice(
+        subkey1,
+        num_steps,
+        shape=(images_train.shape[0],),
+        p=t_weights/t_weights.sum()
+    )
+
+    # Generate Gaussian noise for current batch
+    noise = jax.random.normal(subkey2, images_train.shape)
+
+    # Execute training step with noise prediction and parameter updates
+    loss = train_step(
+        model, optimizer, images_train, t, noise,
+        diffusion.sqrt_alpha_cumulative, diffusion.sqrt_one_minus_alpha_cumulative
+    )
+
+    # Update exponential moving average of loss for smoother tracking
+    if moving_avg_loss is None:
+        moving_avg_loss = loss
+    else:
+        moving_avg_loss = beta * moving_avg_loss + (1 - beta) * loss
+
+    losses.append(moving_avg_loss)
+
+    # Log training progress at regular intervals
+    if epoch % 100 == 0:
+        print(f"Epoch {epoch}, Loss: {moving_avg_loss:.4f}")
+
+print("\nTraining completed.")
+```
+
++++ {"id": "SHX69WgxKUs7"}
+
+
+
++++ {"id": "s-iqch4HKlBV"}
+
+## Training loss visualization
+
+It uses a logarithmic scale to better display the exponential decay of the loss values over time. This representation helps identify both early rapid improvements and later fine-tuning phases of the training process.
+
+```{code-cell}
+---
+colab:
+  base_uri: https://localhost:8080/
+  height: 487
+id: 1bjvWNCcbN24
+outputId: 457fd13f-377f-4021-ddc2-e36940b42550
+---
+# Plot the training loss history with logarithmic scaling
+plt.figure(figsize=(10, 5))            # Create figure with wide aspect ratio for clarity
+plt.plot(losses)                       # losses: List[float] - historical EMA loss values
+plt.title('Training Loss Over Time')
+plt.xlabel('Epoch')
+plt.ylabel('Loss')
+plt.yscale('log')                      # Use log scale to better visualize exponential decay
+plt.grid(True)                         # Add grid for easier value reading
+plt.show()
+```
+
++++ {"id": "M2ql0KwYJLqn"}
+
+## Visualization Functions
+
+Includes utilities for:
+- Sample generation
+- Forward/reverse process visualization
+- Training progress tracking
```

Comment:
```suggestion
In this section, we include utilities for:
- Sample generation
- Forward/reverse process visualization

These tools can help you track the training progress for your model.
```

I didn't see a "Training progress tracking" code cell but maybe you mean something different, such as saying that the sample generation and forward/reverse process visualization help you track the training progress. In that case I would spell it out and not add it as a bullet item.

Reply:
I'm adjusting part of it - As in the common Google tech writing guideline, I do prefer to keep it without the "We language".